pennylane/templates/base/single.py

```diff
@@ -0,0 +1,240 @@
+# Copyright 2018-2019 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+r"""
+Contains the ``Single`` base template.
+"""
+#pylint: disable-msg=too-many-branches,too-many-arguments,protected-access
+from collections import Iterable
+
+from pennylane.templates.decorator import template
+from pennylane.templates.utils import (_check_wires,
+                                       _check_type,
+                                       _get_shape)
+
+
+@template
+def Single(unitary, wires, parameters=None, kwargs={}):
```

Comment:
I was first leaning towards suggesting replacing `unitary` with `operation` because for me "unitary" makes me think I can put in a unitary matrix. But then I saw that one can also pass templates as arguments. And I actually can't come up with a better word that includes "operation" and "template". "template" would be a candidate, because single operations are also templates, abstractly speaking, but that is not so clear to the user.

Reply:
I recommend to replace the name `Single` with `map`, `apply` or something along this line, because it describes better what is actually happening.

Reply:
How about just having the first argument be `template`? That makes the most sense to me (since this is essentially 'mapping' a template/operation over wires).

With regards to names, we are already using `map` and `apply` for mapping/applying templates to measurement outputs, so that might be confusing (although I agree it does tend to make sense).

Is there a NumPy equivalent we can use? How about [`broadcast`](https://docs.scipy.org/doc/numpy/user/basics.broadcasting.html)?

_Note:_ since this is not strictly a template, but a function for constructing templates, it should have a lower case name.

Reply:
Thoughts?
```python
@qml.template
def layer(x, y, wires):
    qml.Hadamard(wires=wires[0])
    qml.RX(x, wires=wires[0])
    qml.RY(y, wires=wires[0])

@qml.qnode(dev)
def circuit(x):
    qml.broadcast(qml.RX, wires=[0, 1, 2], parameters=[0.2, 0.1, 0.4])
    qml.broadcast(layer, wires=[0, 2], parameters=[[0.2, 0.1], [0.4, 0.1]]) # are params nested here(?)
```

Reply:
`broadcast` is fine, or maybe `map_template`

Reply:
Wait...this is just the first base template out of many. There will be all sorts of architectures

Reply:
Also, this is not only a building block, but could be called as a "standalone" template, so it should *definitely* be upper case...

Reply:
Changing the name ``unitary`` to ``template`` is very misleading if feeding a gate only. I thought previously something like ``base_unitary`` or ``block`` could work, but I am not sure if ``template`` is a good idea.

Reply:
block sounds cool

Reply:
refactored to block